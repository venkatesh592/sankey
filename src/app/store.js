import { configureStore } from '@reduxjs/toolkit';
import sankeyReducer from '../features/sankey/sankeySlice';

export default configureStore({
  reducer: {
    options: sankeyReducer,
  },
});
