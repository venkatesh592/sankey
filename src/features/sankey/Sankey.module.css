.sankey {
  width: 100vw;
  height: 100vh;
}

.container {
  display: flex;
}

.usage {
  display: flex;
  align-items: flex-start;
  text-align: left;
  line-height: 1.6;
  padding: 15px;
}