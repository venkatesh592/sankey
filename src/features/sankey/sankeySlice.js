import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  title: {
    left: 'center'
  },
  backgroundColor: "#FFFFFF",
  series: [
    {
      type: "sankey",
      left: 50.0,
      top: 20.0,
      right: 150.0,
      bottom: 25.0,
      data: [],
      links: [],
      lineStyle: {
        color: "source",
        curveness: 0.5
      },
      itemStyle: {
        color: "#1f77b4",
        borderColor: "#1f77b4"
      },
      label: {
        color: "rgba(0,0,0,0.7)",
        fontFamily: "Arial",
        fontSize: 10
      }
    }],
  tooltip: {
    trigger: "item"
  }
}

const getDataAndLinks = (state) => {
  const { series: [ firstSeries ]  } = state
  return firstSeries
}

export const sankeySlice = createSlice({
  name: 'options',
  initialState,
  reducers: {
    setSeriesData: (state, action) => {
      const { data } = getDataAndLinks(state)

      // Avoid the duplicate
      const isExists = data.find(point => point.name === action.payload.name)
      if (!isExists) {
        data.push(action.payload)
      }
    },
    setSeriesLink: (state, action) => {
      const { links } = getDataAndLinks(state)
      links.push(action.payload)
    },
    updateDataPointName: (state, { payload: { index, name } }) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      const { data, links } = getDataAndLinks(state)
      // Avoid duplicates
      const isExists = data.find(point => point.name === name)
      if (isExists) {
        return
      }
      const oldName = data[index].name
      data[index].name = name
      const filteredLinks = links.filter(link => (link.source === oldName || link.target === oldName))
      filteredLinks.forEach(link => {
        link.source = link.source === oldName ? name : link.source
        link.target = link.target === oldName ? name : link.target
      });
    },
    removeDataPoint: (state, { payload: { index } }) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      const { data, links } = getDataAndLinks(state)
      const oldName = data[index].name
      const filteredLinks = links.filter(link => (link.source === oldName || link.target === oldName))
      filteredLinks.forEach((link, index) => {
        links.splice(index, 1)
      })
      data.splice(index, 1)
    },
    removeDataLink: (state, { payload: { index } }) => {
      const { links } = getDataAndLinks(state)
      links.splice(index, 1)
    }
  },
});

export const { setSeriesData, setSeriesLink, updateDataPointName, removeDataPoint, removeDataLink } = sankeySlice.actions;

export const selectDataPoints = state => getDataAndLinks(state.options).data;
export const selectDataLinks = state => getDataAndLinks(state.options).links;
export const selectOptions = state => state.options;

export default sankeySlice.reducer;
