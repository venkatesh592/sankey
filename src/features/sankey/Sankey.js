import React from 'react';

import 'echarts/lib/chart/sankey';
import 'echarts/lib/component/title';
import 'echarts/lib/component/tooltip'

import ReactEcharts from 'echarts-for-react';
import styles from './Sankey.module.css';
import { selectOptions } from './sankeySlice';
import { useSelector } from 'react-redux';

import {
  DataLink,
  DataPoint,
  DataPointList,
  DataLinkList
} from '../index';

export function Sankey() {
  const sankeyOption = useSelector(selectOptions)

  return (
    <div className={styles.sankey}>
      <div className={styles.usage}>
        Usage: (Feed the Data)
        <br />
        1) Create Data point (should be unique)
        <br />
        2) Create Link between them
        <br />
        Example: Salary > Home [3000] > [Electricity - 1000, Grocery - 2000]
        <br />
        Steps:
        <br />
        1) Create four data points with names like [Salary, Home, Electricity, Grocery]
        <br />
        2) Create three links between them 
        <br />
        a) source: Salary, target: Home, Expenditure: 3000
        b) source: Home, target: Electricity, Expenditure: 1000
        c) source: Home, target: Grocery, Expenditure: 2000
        <br />
        You can see graph below based on inputs. (you can also do edit and delete options as well)
        <br />
        Note: Description only availalebe in English
      </div>
      <div className={styles.container}>
        <DataPoint></DataPoint>
        <DataLink></DataLink>
      </div>
      <div className={styles.container}>
        <DataPointList></DataPointList>
        <DataLinkList></DataLinkList>
      </div>
      <ReactEcharts
        style={{height: '100%', width: '100%'}}
        option={sankeyOption} />
    </div>
  )
}
