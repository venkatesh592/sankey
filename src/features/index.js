import { DataLink } from './data-link/DataLink'
import { DataLinkList } from './data-link-list/DataLinkList'
import { DataPoint } from './data-point/DataPoint'
import { DataPointList } from './data-point-list/DataPointList'
import { EditDialog } from './dialog/EditDialog'
import { Header } from './header/Header'
import { Sankey } from './sankey/Sankey'

export {
  DataLink,
  DataLinkList,
  DataPoint,
  DataPointList,
  EditDialog,
  Header,
  Sankey
}
