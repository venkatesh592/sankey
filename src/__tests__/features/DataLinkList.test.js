import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../app/store';
import { DataLinkList } from '../../features';
import { setSeriesLink } from '../../features/sankey/sankeySlice';

let container = null
beforeEach(async () => {
  let link = {
    source: 'usage',
    target: 'usage 2',
    total: 123
  }
  await store.dispatch(setSeriesLink(link))
  container = render(
    <Provider store={store}>
      <DataLinkList />
    </Provider>
  );
})

test('snapshot', () => {
  expect(container.asFragment()).toMatchSnapshot();
});

afterAll(() => {
  jest.clearAllMocks()
})
