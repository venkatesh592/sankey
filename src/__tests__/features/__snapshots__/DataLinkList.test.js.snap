// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav
      aria-label="main mailbox folders"
      class="MuiList-root MuiList-padding"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            usage usage 2
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </li>
    </nav>
  </div>
</DocumentFragment>
`;
